bagua/torch_api/distributed.py

```diff
@@ -93,23 +98,100 @@ def produces_sparse_gradient(module):
+                            p.grad = p.grad.to_sparse()
+            optimizer.step()
+            optimizer_state_dict = optimizer.state_dict()
+        if len(optimizer_state_dict["state"]) == 0:
+            return
+
+        def _state_param_callback(param_id, param_name):
+            def _assign_state(v):
+                optimizer_state_dict["state"][param_id][param_name] = v
+            return _assign_state
+
+        def _hyper_param_callback(index, group_key):
+            def _assign_hyper(v):
+                optimizer.param_groups[index][group_key] = v
+            return _assign_hyper
+
+        params = []
+        scalars = collections.OrderedDict()
+        call_back_param = {}
+        repeat_param_count = collections.defaultdict(int)
+
+        # All "sorted()" operations in this module are used to
+        # guarteen the scalar's record order samely in differet ranks.
+        for index, param_group in enumerate(optimizer_state_dict["param_groups"]):
+            for group_key, group_value in sorted(
+                param_group.items(), key=lambda item: item[0]
+            ):
+                # Hyper-parameters like learning rate are scalars, we need to broadcast them separately.
+                if group_key != "params":
+                    key = "%s_%d" % (group_key, index)
+                    scalars[key] = group_value
+                    call_back_param[key] = _hyper_param_callback(index, group_key)
+            for param_id in sorted(param_group["params"]):
+                if param_id not in optimizer_state_dict["state"]:
+                    continue
+                param_state = optimizer_state_dict["state"][param_id]
+                for param_name, inner_state in sorted(
+                    param_state.items(), key=lambda item: item[0]
+                ):
+                    # Some parameter names, e.g., step, may appear more than once, in which
+                    # case we ensure they have a unique identifier defined by
+                    # their order.
+                    repeat_param_count[param_name] += 1
+                    key = "%s_%d" % (str(param_name), repeat_param_count[param_name])
+                    if isinstance(inner_state, torch.Tensor):
+                        params.append((key, inner_state))
+                    else:
+                        scalars[key] = inner_state
+                        call_back_param[key] = _state_param_callback(
+                            param_id, param_name
+                        )
+        for key, param in params:
+            broadcast(param, src=0)
+        scalars = self._bagua_broadcast_scalars(scalars, src=0)
+        for key, p in scalars.items():
+            call_back_param[key](p)
+
+    def _bagua_broadcast_scalars(self, scalars, src):
+        # Serializes and broadcast scalars by converting them to "ByteTensor".
+        b = io.BytesIO()
+        cloudpickle.dump(scalars, b)
```

Comment:
It seems pickle can do the same thing.

Reply:
According to the 2nd paragraph of https://github.com/cloudpipe/cloudpickle#overriding-pickles-serialization-mechanism-for-importable-constructs, cloudpickle seems more suitable for distributed execution environment. Maybe more safe？